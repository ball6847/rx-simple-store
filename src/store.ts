import { BehaviorSubject } from "rxjs/BehaviorSubject";
import { Observable } from "rxjs/Observable";

export interface RxStoreOptions {
  debug?: boolean;
  mutateState?: boolean;
}

/**
 * This abstract store is basically just holding a state and propagating state changes.
 * Whether this state is treated mutable or immutable is up to you.
 */
export abstract class RxStore<S extends object> {
  readonly state$: Observable<S>;
  private stateSubject: BehaviorSubject<S>;

  get state() {
    return this.stateSubject.value;
  }

  constructor(initialState: S, private options: RxStoreOptions = {}) {
    this.stateSubject = new BehaviorSubject(initialState);
    this.state$ = this.stateSubject.asObservable();
  }

  /**
   * You should use this method to propagate state changes to your application.
   * There are two ways to use this method:
   * 1. The mutable way:
   *   You apply changes to the state object and then call this method without
   *   parameters.
   * 2. The immutable way:
   *   You pass in a partial state which reflects the state changes.
   * @param patch This partial state will be merged into the new resulting state.
   */
  protected propagate(patch: Partial<S>) {
    // tslint:disable-next-line:prefer-object-spread
    const state = this.options.mutateState
      ? Object.assign(this.state, patch)
      : Object.assign({}, this.state, patch);
    if (this.options.debug) {
      // tslint:disable-next-line:no-console
      console.debug(`State changed [${this.constructor.name}]:`, state);
    }
    this.stateSubject.next(state);
  }
}
