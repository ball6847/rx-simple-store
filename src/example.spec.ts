import { RxStore } from "./store";

interface Todo {
  text: string;
  done: boolean;
}

interface State {
  todos: Todo[];
}

class TodoStore extends RxStore<State> {
  // This action is not mutating the state
  addTodo(todo: Todo) {
    this.publish({
      todos: [...this.state.todos, todo],
    });
  }

  // This action is mutating the state
  updateTodo(idx: number, todo: Partial<Todo>) {
    const todos = this.state.todos;
    this.publish({
      todos: [
        ...todos.slice(0, idx),
        Object.assign(todos[idx], todo),
        ...todos.slice(idx + 1),
      ],
    });
  }
}

const todoStore = new TodoStore({ todos: [] });

todoStore.addTodo({ text: "Read the docs", done: false });
todoStore.updateTodo(0, { done: true });
todoStore.addTodo({ text: "Use it", done: false });
todoStore.updateTodo(1, { text: "Apply it" });

describe("Example", () => {
  it("should result in the correct state", () => {
    expect(todoStore.state).toMatchSnapshot();
  });
});
