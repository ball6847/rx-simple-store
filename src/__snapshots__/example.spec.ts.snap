// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example should result in the correct state 1`] = `
Object {
  "todos": Array [
    Object {
      "done": true,
      "text": "Read the docs",
    },
    Object {
      "done": false,
      "text": "Apply it",
    },
  ],
}
`;
